lib/src/path_parser.cc

```diff
@@ -0,0 +1,251 @@
+                                                             simple_config_origin origin,
+                                                             string const& original_text,
+                                                             config_syntax flavor)
+    {
+        token_list tokens;
+        path path = parse_path_expression(expression, origin, original_text, &tokens, flavor);
+        return config_node_path(path, tokens);
+    }
+
+    path path_parser::parse_path_expression(token_iterator& expression, simple_config_origin origin,
+                                            string const& original_text, token_list* path_tokens,
+                                            config_syntax flavor)
+    {
+        vector<element> elements = { element("", false) };
+
+        if (!expression.has_next()) {
+            throw config_exception("Invalid path: " + original_text +
+                                           " -- Expecting a field name or path here, but got nothing");
+        }
+
+        while (expression.has_next()) {
+            shared_token t = expression.next();
+
+            if (path_tokens != nullptr) {
+                path_tokens->push_back(t);
+            }
+
+            if (t->get_token_type() == token_type::IGNORED_WHITESPACE) {
+                continue;
+            }
+
+            if (tokens::is_value_with_type(t, config_value_type::STRING)) {
+                auto value = tokens::get_value(t);
+                // this is a quoted string, so any periods in here don't count as path separators
+                string s = value->transform_to_string();
+                add_path_text(elements, true, s);
+            } else if (t == tokens::end_token()) {
+                // ignore this; when parsing a file, it should not happen
+                // since we're parsing a token list rather than the main
+                // token iterator, and when parsing a path expression from the
+                // API, it's expected to have an END.
+            } else {
+                // any periods outside of a quoted string count as separators
+                string text;
+                if (t->get_token_type() == token_type::VALUE) {
+                    // appending a number here may add a period, but we _do_ count those as path
+                    // separators, because we basically want "foo 3.0bar" to parse as a string even
+                    // though there's a number in it. The fact that we tokenize non-string values
+                    // is largely an implementation detail.
+                    auto value = tokens::get_value(t);
+
+                    // We need to split the tokens on a . so that we can get sub-paths but still preserve
+                    // the original path text when doing an insertion
+                    if (path_tokens != nullptr) {
+                        path_tokens->pop_back();
+                        for (auto&& token : split_token_on_period(t, flavor)) {
```

Comment:
This could use `path_tokens->insert`. The only reason I prefer that is to avoid having to reallocate the token_list when we already know how many elements we're adding.